File: 2024/day2/main.clj
#!/usr/bin/env clojure

;; The task: Each row in input is a report and each value in each report is
;;a level. SAFE reports are ones that are either all increasing or all
;;decreasing, AND the difference between two adjacent levels are AT MOST 3
;; How many SAFE reports are in this input?

(ns main
  (:require [clojure.string :as str])
  (:gen-class))

;; extract each report into vector
(def input-reports
  (->> (slurp "./input.txt")
       (str/split-lines)
       (map #(str/split % #" "))
       (map #(map parse-long %))))

;;;;;;;;;;;;;;;;;
;     PART1     ;
;;;;;;;;;;;;;;;;

;; If all records are either increasing or decreasing
(defn monotonic-diffs? [diffs]
  (or (every? pos? diffs) (every? neg? diffs)))

;; If all diffs are within the specified bounds
(defn diffs-within-bounds? [diffs]
  (every? #(and (<= (abs %) 3) (>= (abs %) 1)) diffs))

;; Check if report is safe
;; 1. The levels are either all increasing or all decreasing.
;; 2. Any two adjacent levels differ by at least one and at most three.
(defn safe-report? [report]
  ; get the diff of 1st element with 2nd, 3rd with 4th and so on...
  (let [diffs (map #(- %2 %1) report (rest report))]
    (and (monotonic-diffs? diffs) (diffs-within-bounds? diffs))))

;; count the safe reports
;; Accepts the function which safe reports are calculated with, and the list of reports
(defn count-safe-reports [function reports]
  (count (filter function reports)))

;;;;;;;;;;;;;;;;;
;     PART2     ;
;;;;;;;;;;;;;;;;
;; Part2 is the same as part1 but now we have tolerance for one bad level.

;; METHOD:
;; Check if report is safe like PART1
;; If report is unsafe, start removing each element. if it becomes safe by removing an element,
;;The report is safe. if not, the report is not safe
(defn safe-report-with-removal? [report]
  (if (safe-report? report)
    ;if report is already safe
    true
    ; Remove elements one by one and see if report becomes safe
    (some (fn [index]
            ; create report-new without the element at index
            (let [report-new (vec (concat (take index report) (drop (inc index) report)))]
              (safe-report? report-new)))
          ; create lazy seq from 0 to (count report) to use as index of predicate
          (range (count report)))))

(println (count-safe-reports safe-report? input-reports))
(println (count-safe-reports safe-report-with-removal? input-reports))

File: 2024/day2/main.js
const REPORTS_OK = [
    [1, 2, 4, 3, 5],
    [1, 3, 4, 7],
    [4, 1, 5, 6],
    [6, 2, 7, 9],
    [9, 7, 5, 2, 3]
];

const REPORTS_BAD = [
    [10, 11, 8, 7, 3],
    [6, 2, 7, 9, 8],
    [1, 3, 5, 4, 7, 7]
];

const inputReports = Deno.readTextFileSync('./input.txt', 'utf-8')
    .split('\n') // Split the content into lines
    .filter(line => line.trim() !== '') // Filter out empty lines
    .map(line => line.split(' ')) // Split each line into words
    .map(arr => arr.map(Number)); // Convert each word to a number

function countSafeReports() {
    return inputReports.filter(isReportSafe).length
}

function countSafeReportsWithRemoval() {
    return inputReports.filter(isReportSafeWithRemoval).length
}

/**
 * Check if a given report is safe.
 *
 * For a report to be safe, it must:
 * - The levels are either all increasing or all decreasing.
 * - Any two adjacent levels differ by at least one and at most three.
 *
 * @param {number[]} report
 */
function isReportSafe(report) {
    // calculate report[i+1] - report[i]
    const diffs = getDiffs(report);

    // if all records are not either increasing or decreasing
    if (!(diffs.every(x => x > 0) || diffs.every(x => x < 0))) {
        return false;
    }

    for (let i = 0; i < diffs.length; i++) {
        const d = diffs[i];

        if (Math.abs(d) > 3 || Math.abs(d) < 1) {
            return false;
        }
    }

    return true;
}

/**
 * is the report safe if a level is removed?
 *
 * Complexity of this solution is O(n^m) where n is the length of the report and m is the number of
 * faults to tolerate +1
 *
 * @param {number[]} report
 */
function isReportSafeWithRemoval(report) {
    // If the report is already safe
    if (isReportSafe(report)) {
        return true;
    }

    // check by removing one element at a time
    for (let i = 0; i < report.length; i++) {
        // Remove the i-th element from the report
        let reportNew = [...report];
        reportNew.splice(i, 1);
        // console.log(reportNew);
        // see if it is safe after the removal
        if (isReportSafe(reportNew)) {
            return true;
        }
    }

    return false;
}

/**
* Get diff of every level with it's next level
* @param {number[]} value
* @returns {number[]}
*/
function getDiffs(value) {
    let diffs = [];

    for (let i = 0; i < value.length - 1; i++) {
        diffs[i] = value[i + 1] - value[i];
    }

    return diffs
}
